### ISSUE TITLE
Support Promise async calls

### ISSUE THREAD

**Issue by compulim** [l12s/node-slack-client#15]
As Promise interface is going mainstream in ES6, I think `node-slack-client` should also support both callback and Promise on all async calls.

To bridge between callback and Promise, async functions should also return a Promise object, in addition to accepting a callback function.

For example, `readFile` used to return `undefined`. To support Promise, `readFile` will return a Promise object so the caller can use `.then()` to chain their async operations. In the old days,

```javascript
readFile('hello.txt', function (err, buffer) {
  ...
});
```

To support ES6, we also return Promise object in `readFile` call, so we can use `.then` to chain async operations.

```javascript
readFile('hello.txt')
  .then(function (buffer) {
  });
```

According to http://kangax.github.io/compat-table/es6/, basic Promise support is already supported in Node.js 0.12 and modern browsers.

For browser which do not support Promise natively, there are few good polyfills. One is [q](https://www.npmjs.com/package/q), another is [bluebird](https://www.npmjs.com/package/bluebird).

**Comment by ekmartin** [comment]
Looks like that's the idea: https://github.com/l12s/node-slack-client/blob/master/lib%2Fclients%2Fweb%2Ffacets%2Fchat.js#L52

This could be done quite cleanly if Bluebird was used as the Promise implementation, through `asCallback`: http://bluebirdjs.com/docs/api/ascallback.html - also works the opposite direction, i.e. by creating a promise from a callback implementation using promisification. The first is probably preferred though, i.e. having the node-slack-client codebase use promises underneath, and provide callbacks as a measure towards backwards compatibility.

**Comment by compulim** [comment]
BTW, currently, callback is not optionally. I tried `postMessage` without giving a callback, got an uncaught exception.

**Comment by ekmartin** [comment]
Yep @compulim, made an issue for it earlier today :smiley: - #16

**Comment by l12s** [comment]
Yep, I originally implemented this to support promises, but decided to hold off on that for now as some of the retry logic got a little involved.

What I'd suggest is that for the v2.0.0 release you use bluebird's promisify API to make the callback functions support Promises. Once I've got v2.0.0 released and figured out what migration looks like, we can talk about what to on promises.

I've fixed the non-optional cb issue now.
